File: Source/Java/Expressive/src/main/java/com/bijington/expressive/functions/mathematical/CountFunction.java
package com.bijington.expressive.functions.mathematical;

import com.bijington.expressive.exceptions.ExpressiveException;
import com.bijington.expressive.expressions.IExpression;
import com.bijington.expressive.functions.FunctionBase;

import java.lang.reflect.Array;

/**
 * Created by shaun on 18/07/2016.
 */
public class CountFunction extends FunctionBase {
    @Override
    public String getName() {
        return "Count";
    }

    @Override
    public Object evaluate(IExpression[] parameters) throws ExpressiveException {
        this.validateParameterCount(parameters, -1, 1);

        int count = 0;

        for (int parameterIndex = 0; parameterIndex < parameters.length; parameterIndex++) {
            IExpression value = parameters[parameterIndex];

            int increment = 1;
            Object evaluatedValue = value.evaluate(this.getVariables());

            if (evaluatedValue != null &&
                    evaluatedValue.getClass().isArray()) {
                increment = Array.getLength(evaluatedValue);
            }

            count += increment;
        }

        return count;
    }
}

File: Source/CSharp/Expressive/Expressive/Functions/Mathematical/CountFunction.cs
using Expressive.Expressions;
using System;
using System.Collections;

namespace Expressive.Functions.Mathematical
{
    internal class CountFunction : FunctionBase
    {
        #region FunctionBase Members

        public override string Name { get { return "Count"; } }

        public override object Evaluate(IExpression[] parameters, ExpressiveOptions options)
        {
            this.ValidateParameterCount(parameters, -1, 1);

            int count = 0;

            foreach (var value in parameters)
            {
                int increment = 1;
                object evaluatedValue = value.Evaluate(Variables);
                IEnumerable enumerable = evaluatedValue as IEnumerable;

                if (enumerable != null)
                {
                    int enumerableCount = 0;
                    foreach (var item in enumerable)
                    {
                        enumerableCount++;
                    }
                    
                    increment = enumerableCount;
                }

                count += increment;
            }

            return count;
        }

        #endregion
    }
}
